File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.java
class Solution {

    public boolean same(TreeNode a, TreeNode b) {
        if (a == null && b == null) return true;
        if (a == null || b == null) return false;

        if (a.val != b.val) return false;

        return same(a.left, b.left) &&
               same(a.right, b.right);
    }

    public boolean isSubtree(TreeNode root, TreeNode subRoot) {

        if (subRoot == null) return true;
        if (root == null) return false;

        if (same(root, subRoot)) return true;

        return isSubtree(root.left, subRoot) ||
               isSubtree(root.right, subRoot);
    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if root is None or p is None or q is None:
            return None
        
        temp=root
        while temp is not None:
            if temp.val>p.val and temp.val>q.val:
                temp=temp.left
            elif temp.val<p.val and temp.val<q.val:
                temp=temp.right
            else:
                return temp